modules/51DegreesRtdProvider.js

```diff
@@ -0,0 +1,255 @@
+ *
+ * @param {Object} obj The object to set the key in
+ * @param {string} key The key to set
+ * @param {any} value The value to set
+ */
+export const setOrtb2KeyIfNotEmpty = (obj, key, value) => {
+  if (!key) {
+    throw new Error(LOG_PREFIX + ' Key is required');
+  }
+
+  if (value) {
+    obj[key] = value;
+  }
+}
+
+/**
+ * Converts 51Degrees device data to ORTB2 format
+ *
+ * @param {Object} device
+ * @param {string} [device.deviceid] Device ID (unique 51Degrees identifier)
+ * @param {string} [device.devicetype]
+ * @param {string} [device.hardwarevendor]
+ * @param {string} [device.hardwaremodel]
+ * @param {string[]} [device.hardwarename]
+ * @param {string} [device.platformname]
+ * @param {string} [device.platformversion]
+ * @param {number} [device.screenpixelsheight]
+ * @param {number} [device.screenpixelswidth]
+ * @param {number} [device.pixelratio]
+ * @param {number} [device.screeninchesheight]
+ *
+ * @returns {Object}
+ */
+export const convert51DegreesDeviceToOrtb2 = (device) => {
+  const ortb2Device = {};
+
+  if (!device) {
+    return ortb2Device;
+  }
+
+  const deviceModel =
+    device.hardwaremodel || (
+      device.hardwarename && device.hardwarename.length
+        ? device.hardwarename.join(',')
+        : null
+    );
+
+  const devicePPI = device.screenpixelsheight && device.screeninchesheight
+    ? Math.round(device.screenpixelsheight / device.screeninchesheight)
+    : null;
+
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'devicetype', ORTB_DEVICE_TYPE_MAP.get(device.devicetype));
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'make', device.hardwarevendor);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'model', deviceModel);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'os', device.platformname);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'osv', device.platformversion);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'h', device.screenpixelsheight);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'w', device.screenpixelswidth);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'pxratio', device.pixelratio);
+  setOrtb2KeyIfNotEmpty(ortb2Device, 'ppi', devicePPI);
+
+  if (device.deviceid) {
+    ortb2Device.ext = {
+      'fiftyonedegrees_deviceId': device.deviceid
+    };
+  }
+
+  return ortb2Device;
+}
+
+/**
+ * @param {Object} reqBidsConfigObj Bid request configuration object
+ * @param {Function} callback Called on completion
+ * @param {Object} moduleConfig Configuration for 1plusX RTD module
+ * @param {Object} userConsent
+ */
+export const getBidRequestData = (reqBidsConfigObj, callback, moduleConfig, userConsent) => {
+  try {
+    // Get the required config
+    const {resourceKey, onPremiseJSUrl} = extractConfig(moduleConfig, reqBidsConfigObj);
+    logMessage('Resource key: ', resourceKey);
+    logMessage('On-premise JS URL: ', onPremiseJSUrl);
+
+    // Get 51Degrees JS URL, which is either cloud or on-premise
+    const scriptURL = get51DegreesJSURL(resourceKey ? {resourceKey} : {onPremiseJSUrl});
```

Comment:
If you're using prebid to inject, we prefer you lock versions and pr the project when the script changes

Reply:
The script is generated by the [51Degrees Cloud API](https://cloud.51degrees.com/api-docs/index.html) that is version-locked through its URL:  `https://cloud.51degrees.com/api/v4/<your_resource_Key>.js`, however the script itself is generated based on resource-key and properties user selects for it - so we would not be able to store its content as part of this repo.  The javascript template and device detection library used by the cloud are open source (as mentioned in our docs): 
- https://github.com/51Degrees/javascript-templates/
- https://github.com/51Degrees/device-detection-dotnet

Reply:
Can you dm me a copy of something running; I want to check for example, you don't gather bids or identifiers from other companies and send them to your endpoint, or other obvious rules violations.